scripts/get-aidrop-sources.ts

```diff
@@ -0,0 +1,222 @@
+
+import { ethers, upgrades } from "hardhat";
+import { BigNumber, Contract, ContractFactory } from "ethers";
+import Axios, { AxiosInstance } from "axios";
+import fs from "fs";
+import path from "path";
+const cliProgress = require('cli-progress');
+const _colors = require('colors');
+
+const alchemyUrl = 'https://eth-mainnet.alchemyapi.io/v2/6Kv_bJKux1FXeX5h7ga3GaxICthIx9UG';
+const etherscanUrl = 'https://api.etherscan.io/api';
+
+const tokenAddress = "0x0C63cAE5fcC2Ca3dDE60a35e50362220651eBEc8";
+const UNISWAP_stXEM = '0xab2f9b13b4aca6884fdbb9e4a1f767c82c53688f'
+const poolAddresses: string[] = [UNISWAP_stXEM];
+const blockNumber = undefined;
+
+async function getTokenBalance(
+  data: string,
+  contractAddress: string,
+  blockNumber?: number
+): Promise<BigNumber> {
+  const alchemyApi = Axios.create({
+    baseURL: alchemyUrl,
+    headers: {},
+    timeout: 120000,
+  });
+
+  const params: any = [{
+    data,
+    "to": contractAddress
+  }]
+
+  if (blockNumber) {
+    params.push('0x' + blockNumber.toString(16))
+  }
+
+  const result = await alchemyApi.post('', {
+    "id": 1,
+    "jsonrpc": "2.0",
+    "method": "eth_call",
+    params,
+  })
+
+  if (!result || !result.data || result.data.result === '0x') {
+    return ethers.BigNumber.from(0);
+  }
+
+  return result.data.result;
+}
+
+
+async function getTokenLogs(address: string): Promise<any[]> {
```

Comment:
What's the reason to query logs from etherscan instead of directly from the token using ethers.js like this?

```
const StakedToken: ContractFactory = await ethers.getContractFactory("StakedToken");
  const stakedToken = StakedToken.attach(stakedTokenAddress) as StakedToken;

  const filter = stakedToken.filters.Transfer(null, null, null, null);
  const transfers = await stakedToken.queryFilter(filter, 0);
  ```

Reply:
In which case I think you wouldn't have to parse them manually either

Reply:
@MuncleUscles i checked this, and by looking only for Transfer events we don't get all the addresses. On the other hand, when calling the smart contract directly, we can get specific events and add filters. So it seems we would have to call smart contract for every single event. Instead by using the API, we get all the events in one call.